ldap/client/api/src/main/java/org/apache/directory/ldap/client/api/LdapNetworkConnection.java

```diff
@@ -599,102 +604,100 @@ public boolean connect() throws LdapException
-                        continue;
-                    }
+            if ( result.getConnectFuture() == null || !result.getConnectFuture().isConnected() )
+            {
+                result.getConnector().dispose();
+                continue;
+            }
 
-                    if ( futureResult == null || !futureResult.isConnected() )
-                    {
-                        cf.getConnector().dispose();
-                        continue;
-                    }
+            connectionFuture = result.getConnectFuture();
+            connector = result.getConnector();
 
-                    connFuture = cf;
-                    connectionFuture = futureResult;
-                    setConnector( cf.getConnector() );
+            break;
+        }
 
-                    break breakaway;
-                }
+        for ( Future<LdapNetworkConnectionFuture> future : futures )
+        {
+            LdapNetworkConnectionFuture futureToCancel;
+            try
+            {
+                futureToCancel = future.get();
             }
-
-            for ( LdapNetworkConnectionFuture cf : connFutures )
+            catch ( InterruptedException e )
             {
-                try
-                {
-                    if ( cf.getConnectFuture().isDone() && cf.getConnectFuture().get().isConnected() )
-                    {
-                        cf.getConnectFuture().get().getSession().closeNow();
-                    }
-                }
-                catch ( InterruptedException e )
-                {
-                    LOG.debug( "InterruptedException thrown while trying to close an open session: {}", e.getMessage() );
-                }
-                catch ( ExecutionException e )
-                {
-                    LOG.debug( "ExecutionException thrown while trying to close an open session: {}", e.getMessage() );
-                }
+                LOG.info( "InterruptedException thrown while cleaning up leftover futures/connections" );
+                future.cancel( true );
+                continue;
+            }
+            catch ( ExecutionException e )
+            {
+                LOG.info( "ExecutionException thrown while cleaning up leftover futures/connections" );
+                future.cancel( true );
+                continue;
+            }
 
-                cf.getConnectFuture().cancel( true );
-                cf.getConnector().dispose();
+            if ( futureToCancel.getConnectFuture().isDone() && futureToCancel.getConnectFuture().isConnected() )
+            {
+                futureToCancel.getConnectFuture().getSession().closeNow();
             }
+
+            futureToCancel.getConnector().dispose();
+            future.cancel( true );
         }
-        executor.shutdown();
 
-        if ( connFuture == null )
+        if ( connectionFuture == null )
         {
             throw new InvalidConnectionException( "Exhausted all options while trying to connect to ldap - giving up!" );
```

Comment:
That's the thing. If it reaches that place, it might be lots of different errors. How should we know what is the proper error to display if there might be many different errors if the system is trying to connect to different ips?

So either make it generic as it is now or print out all the detailed errors that happened.